src/compose.ts

```diff
@@ -346,3 +381,128 @@ function normalizeKeyValuePairs(
 		.fromPairs()
 		.value();
 }
+
+/**
+ * Reads a normalized composition object and expands context of composition
+ * Context is related to the composition in terms of env_file tag in composition
+ * env_file tag values (file paths) are normalized and validated.
+ *
+ * Caches environment files for all services in the composition and transfers all env_file
+ * environment vars into the environment tag of the composition.
+ *
+ * Drops env_file propertie from composition to indicate that the expand has taken place
+ * 
+ * @param c normalized composition object
+ * @param fileResolverCb callback from caller that returns a readStream for a given filePath string
+ */
+async function expandContext(
+	c: Composition,
+	fileResolverCb: (path: string) => Promise<ReadStream>,
+): Promise<Composition> {
+	const cachedEnvironmentFiles: Dict<Dict<string>> = {};
+
+	// Apply all read env_files content to the services referncing the env_files
+	for (const service of Object.values(c.services)) {
+		let envFilePaths = service.env_file;
+		if (!!envFilePaths) {
+			service.environment = service.environment ?? {};
+			if (!Array.isArray(envFilePaths)) {
+				envFilePaths = [envFilePaths];
+			}
+			for (const envFilePath of envFilePaths) {
+				const validatedPath = validateAndNormalizeFilePath(envFilePath);
+				if (!(validatedPath in cachedEnvironmentFiles)) {
+					cachedEnvironmentFiles[validatedPath] = await readAndExpandEnvFile(
+						validatedPath,
+						fileResolverCb,
+					);
+				}
+				for (const [key, value] of Object.entries(
+					cachedEnvironmentFiles[validatedPath],
+				)) {
+					if (!(key in service.environment)) {
+						service.environment[key] = value;
+					}
+				}
+			}
+			delete service.env_file;
```

Comment:
sorry, just thinking through... why are we deleting this?

Reply:
Agree on this, this is a non expected side effect. I've deleted this and updated the test cases

Reply:
@dfunckt I'm not dropping the property as this is an unintended / not necessary manipulation of the input composition.

Reply:
I’m not sure I agree. Why is removing env_file unnecessary manipulation but augmenting “environment” isn’t?

env_file implies a certain processing step which this module takes care of. I don’t see why carrying it over is a better choice.

Reply:
After talking with @dfunckt and @fisehara, we won't worry about argument mutation for now.  This is something to think about in the future.

Reply:
As discussed, this mutation is intended and the env_file property will be deleted. Thanks for discussing this :)